<a href="https://colab.research.google.com/github/Zjmgpk/CS5530/blob/main/Assignment2_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import numpy
import pandas
import sklearn

In [ ]:
df = pandas.read_csv("/content/drive/MyDrive/Colab Notebooks/Assignment 2/Assignment 2_1/Raw_Data/train.csv")
df

Unnamed: 0                              Name    Location  Year  \
0              1  Hyundai Creta 1.6 CRDi SX Option        Pune  2015   
1              2                      Honda Jazz V     Chennai  2011   
2              3                 Maruti Ertiga VDI     Chennai  2012   
3              4   Audi A4 New 2.0 TDI Multitronic  Coimbatore  2013   
4              6            Nissan Micra Diesel XV      Jaipur  2013   
...          ...                               ...         ...   ...   
5842        6014                  Maruti Swift VDI       Delhi  2014   
5843        6015          Hyundai Xcent 1.1 CRDi S      Jaipur  2015   
5844        6016             Mahindra Xylo D4 BSIV      Jaipur  2012   
5845        6017                Maruti Wagon R VXI     Kolkata  2013   
5846        6018             Chevrolet Beat Diesel   Hyderabad  2011   

      Kilometers_Driven Fuel_Type Transmission Owner_Type     Mileage  \
0                 41000    Diesel       Manual      First  19.67 kmpl   
1                 46000    Petrol       Manual      First    13 km/kg   
2                 87000    Diesel       Manual      First  20.77 kmpl   
3                 40670    Diesel    Automatic     Second   15.2 kmpl   
4                 86999    Diesel       Manual      First  23.08 kmpl   
...                 ...       ...          ...        ...         ...   
5842              27365    Diesel       Manual      First   28.4 kmpl   
5843             100000    Diesel       Manual      First   24.4 kmpl   
5844              55000    Diesel       Manual     Second   14.0 kmpl   
5845              46000    Petrol       Manual      First   18.9 kmpl   
5846              47000    Diesel       Manual      First  25.44 kmpl   

       Engine      Power  Seats  New_Price  Price  
0     1582 CC  126.2 bhp    5.0        NaN  12.50  
1     1199 CC   88.7 bhp    5.0  8.61 Lakh   4.50  
2     1248 CC  88.76 bhp    7.0        NaN   6.00  
3     1968 CC  140.8 bhp    5.0        NaN  17.74  
4     1461 CC   63.1 bhp    5.0        NaN   3.50  
...       ...        ...    ...        ...    ...  
5842  1248 CC     74 bhp    5.0  7.88 Lakh   4.75  
5843  1120 CC     71 bhp    5.0        NaN   4.00  
5844  2498 CC    112 bhp    8.0        NaN   2.90  
5845   998 CC   67.1 bhp    5.0        NaN   2.65  
5846   936 CC   57.6 bhp    5.0        NaN   2.50  

[5847 rows x 14 columns]

In [ ]:
df_cleaned = df.dropna(subset=['New_Price'])
display(df_cleaned)

Unnamed: 0                               Name Location  Year  \
1              2                       Honda Jazz V  Chennai  2011   
5              7  Toyota Innova Crysta 2.8 GX AT 8S   Mumbai  2016   
8             10                   Maruti Ciaz Zeta    Kochi  2018   
13            15        Mitsubishi Pajero Sport 4X4    Delhi  2014   
18            20                  BMW 3 Series 320d    Kochi  2014   
...          ...                                ...      ...   ...   
5827        5999              Tata Bolt Revotron XT  Chennai  2016   
5830        6002      Volkswagen Vento 1.6 Highline   Mumbai  2011   
5833        6005           Maruti Vitara Brezza VDi     Pune  2016   
5838        6010               Honda Brio 1.2 VX MT    Delhi  2013   
5842        6014                   Maruti Swift VDI    Delhi  2014   

      Kilometers_Driven Fuel_Type Transmission Owner_Type     Mileage  \
1                 46000    Petrol       Manual      First    13 km/kg   
5                 36000    Diesel    Automatic      First  11.36 kmpl   
8                 25692    Petrol       Manual      First  21.56 kmpl   
13               110000    Diesel       Manual      First   13.5 kmpl   
18                32982    Diesel    Automatic      First  22.69 kmpl   
...                 ...       ...          ...        ...         ...   
5827              10000    Petrol       Manual      First  17.57 kmpl   
5830              38000    Petrol       Manual      First  16.09 kmpl   
5833              37208    Diesel       Manual      First   24.3 kmpl   
5838              33746    Petrol       Manual      First   18.5 kmpl   
5842              27365    Diesel       Manual      First   28.4 kmpl   

       Engine       Power  Seats   New_Price  Price  
1     1199 CC    88.7 bhp    5.0   8.61 Lakh   4.50  
5     2755 CC   171.5 bhp    8.0     21 Lakh  17.50  
8     1462 CC  103.25 bhp    5.0  10.65 Lakh   9.95  
13    2477 CC  175.56 bhp    7.0  32.01 Lakh  15.00  
18    1995 CC     190 bhp    5.0  47.87 Lakh  18.55  
...       ...         ...    ...         ...    ...  
5827  1193 CC    88.7 bhp    5.0   7.77 Lakh   4.00  
5830  1598 CC   103.5 bhp    5.0  11.91 Lakh   3.25  
5833  1248 CC    88.5 bhp    5.0   9.93 Lakh   7.43  
5838  1198 CC    86.8 bhp    5.0   6.63 Lakh   3.20  
5842  1248 CC      74 bhp    5.0   7.88 Lakh   4.75  

[815 rows x 14 columns]

I dropped the NaN of the New_Price column due beacause of how few values there were in the column. With there only being 815 out of 5847, it wouldnt have been possible to fill each of the missing values.

In [ ]:
filtered_df = df_cleaned[(df_cleaned['Kilometers_Driven'] == 50000) & (df_cleaned['Fuel_Type'] == 'Electric')]
display(filtered_df)

Unnamed: 0                  Name Location  Year  Kilometers_Driven  \
4313        4446  Mahindra E Verito D4  Chennai  2016              50000   

     Fuel_Type Transmission Owner_Type Mileage Engine   Power  Seats  \
4313  Electric    Automatic      First     NaN  72 CC  41 bhp    5.0   

       New_Price  Price  
4313  13.58 Lakh   13.0

In [ ]:
rows_with_nan = df_cleaned[df_cleaned.isnull().any(axis=1)]
display(rows_with_nan)

Unnamed: 0                  Name Location  Year  Kilometers_Driven  \
4313        4446  Mahindra E Verito D4  Chennai  2016              50000   

     Fuel_Type Transmission Owner_Type Mileage Engine   Power  Seats  \
4313  Electric    Automatic      First     NaN  72 CC  41 bhp    5.0   

       New_Price  Price  
4313  13.58 Lakh   13.0

In [ ]:
df_cleaned = df_cleaned.drop(4313)
display(df_cleaned)

Name Location  Year  Kilometers_Driven  \
1                          Honda Jazz V  Chennai  2011              46000   
5     Toyota Innova Crysta 2.8 GX AT 8S   Mumbai  2016              36000   
8                      Maruti Ciaz Zeta    Kochi  2018              25692   
13          Mitsubishi Pajero Sport 4X4    Delhi  2014             110000   
18                    BMW 3 Series 320d    Kochi  2014              32982   
...                                 ...      ...   ...                ...   
5827              Tata Bolt Revotron XT  Chennai  2016              10000   
5830      Volkswagen Vento 1.6 Highline   Mumbai  2011              38000   
5833           Maruti Vitara Brezza VDi     Pune  2016              37208   
5838               Honda Brio 1.2 VX MT    Delhi  2013              33746   
5842                   Maruti Swift VDI    Delhi  2014              27365   

     Fuel_Type Transmission Owner_Type     Mileage   Engine       Power  \
1       Petrol    Automatic      First    13 km/kg  1199 CC    88.7 bhp   
5       Diesel    Automatic      First  11.36 kmpl  2755 CC   171.5 bhp   
8       Petrol    Automatic      First  21.56 kmpl  1462 CC  103.25 bhp   
13      Diesel    Automatic      First   13.5 kmpl  2477 CC  175.56 bhp   
18      Diesel    Automatic      First  22.69 kmpl  1995 CC     190 bhp   
...        ...          ...        ...         ...      ...         ...   
5827    Petrol    Automatic      First  17.57 kmpl  1193 CC    88.7 bhp   
5830    Petrol    Automatic      First  16.09 kmpl  1598 CC   103.5 bhp   
5833    Diesel    Automatic      First   24.3 kmpl  1248 CC    88.5 bhp   
5838    Petrol    Automatic      First   18.5 kmpl  1198 CC    86.8 bhp   
5842    Diesel    Automatic      First   28.4 kmpl  1248 CC      74 bhp   

      Seats   New_Price  Price  
1       5.0   8.61 Lakh   4.50  
5       8.0     21 Lakh  17.50  
8       5.0  10.65 Lakh   9.95  
13      7.0  32.01 Lakh  15.00  
18      5.0  47.87 Lakh  18.55  
...     ...         ...    ...  
5827    5.0   7.77 Lakh   4.00  
5830    5.0  11.91 Lakh   3.25  
5833    5.0   9.93 Lakh   7.43  
5838    5.0   6.63 Lakh   3.20  
5842    5.0   7.88 Lakh   4.75  

[814 rows x 13 columns]

Row 4313 was dropped since no other row would relate to it, which made it so I couldnt impute it.

In [ ]:
df_cleaned['Mileage'] = df_cleaned['Mileage'].str.replace(' kmpl', '')
df_cleaned['Mileage'] = df_cleaned['Mileage'].str.replace(' km/kg', '')
df_cleaned['Engine'] = df_cleaned['Engine'].str.replace(' CC', '')
df_cleaned['Power'] = df_cleaned['Power'].str.replace(' bhp', '')
df_cleaned['New_Price'] = df_cleaned['New_Price'].str.replace(' Lakh', '')
df_cleaned['New_Price'] = df_cleaned['New_Price'].str.replace(' Cr', '')
df_cleaned

Name Location  Year  Kilometers_Driven  \
1                          Honda Jazz V  Chennai  2011              46000   
5     Toyota Innova Crysta 2.8 GX AT 8S   Mumbai  2016              36000   
8                      Maruti Ciaz Zeta    Kochi  2018              25692   
13          Mitsubishi Pajero Sport 4X4    Delhi  2014             110000   
18                    BMW 3 Series 320d    Kochi  2014              32982   
...                                 ...      ...   ...                ...   
5827              Tata Bolt Revotron XT  Chennai  2016              10000   
5830      Volkswagen Vento 1.6 Highline   Mumbai  2011              38000   
5833           Maruti Vitara Brezza VDi     Pune  2016              37208   
5838               Honda Brio 1.2 VX MT    Delhi  2013              33746   
5842                   Maruti Swift VDI    Delhi  2014              27365   

     Owner_Type Mileage Engine   Power  Seats New_Price  Price  Diesel  \
1         First      13   1199    88.7    5.0      8.61   4.50   False   
5         First   11.36   2755   171.5    8.0        21  17.50    True   
8         First   21.56   1462  103.25    5.0     10.65   9.95   False   
13        First    13.5   2477  175.56    7.0     32.01  15.00    True   
18        First   22.69   1995     190    5.0     47.87  18.55    True   
...         ...     ...    ...     ...    ...       ...    ...     ...   
5827      First   17.57   1193    88.7    5.0      7.77   4.00   False   
5830      First   16.09   1598   103.5    5.0     11.91   3.25   False   
5833      First    24.3   1248    88.5    5.0      9.93   7.43    True   
5838      First    18.5   1198    86.8    5.0      6.63   3.20   False   
5842      First    28.4   1248      74    5.0      7.88   4.75    True   

      Petrol  Automatic  
1       True       True  
5      False       True  
8       True       True  
13     False       True  
18     False       True  
...      ...        ...  
5827    True       True  
5830    True       True  
5833   False       True  
5838    True       True  
5842   False       True  

[814 rows x 14 columns]

In [ ]:
df_cleaned = pandas.get_dummies(df_cleaned, columns=['Fuel_Type', 'Transmission'], dtype=int)
display(df_cleaned)

Unnamed: 0                               Name Location  Year  \
1              2                       Honda Jazz V  Chennai  2011   
5              7  Toyota Innova Crysta 2.8 GX AT 8S   Mumbai  2016   
8             10                   Maruti Ciaz Zeta    Kochi  2018   
13            15        Mitsubishi Pajero Sport 4X4    Delhi  2014   
18            20                  BMW 3 Series 320d    Kochi  2014   
...          ...                                ...      ...   ...   
5827        5999              Tata Bolt Revotron XT  Chennai  2016   
5830        6002      Volkswagen Vento 1.6 Highline   Mumbai  2011   
5833        6005           Maruti Vitara Brezza VDi     Pune  2016   
5838        6010               Honda Brio 1.2 VX MT    Delhi  2013   
5842        6014                   Maruti Swift VDI    Delhi  2014   

      Kilometers_Driven Owner_Type     Mileage   Engine       Power  Seats  \
1                 46000      First    13 km/kg  1199 CC    88.7 bhp    5.0   
5                 36000      First  11.36 kmpl  2755 CC   171.5 bhp    8.0   
8                 25692      First  21.56 kmpl  1462 CC  103.25 bhp    5.0   
13               110000      First   13.5 kmpl  2477 CC  175.56 bhp    7.0   
18                32982      First  22.69 kmpl  1995 CC     190 bhp    5.0   
...                 ...        ...         ...      ...         ...    ...   
5827              10000      First  17.57 kmpl  1193 CC    88.7 bhp    5.0   
5830              38000      First  16.09 kmpl  1598 CC   103.5 bhp    5.0   
5833              37208      First   24.3 kmpl  1248 CC    88.5 bhp    5.0   
5838              33746      First   18.5 kmpl  1198 CC    86.8 bhp    5.0   
5842              27365      First   28.4 kmpl  1248 CC      74 bhp    5.0   

       New_Price  Price  Fuel_Type_Diesel  Fuel_Type_Electric  \
1      8.61 Lakh   4.50                 0                   0   
5        21 Lakh  17.50                 1                   0   
8     10.65 Lakh   9.95                 0                   0   
13    32.01 Lakh  15.00                 1                   0   
18    47.87 Lakh  18.55                 1                   0   
...          ...    ...               ...                 ...   
5827   7.77 Lakh   4.00                 0                   0   
5830  11.91 Lakh   3.25                 0                   0   
5833   9.93 Lakh   7.43                 1                   0   
5838   6.63 Lakh   3.20                 0                   0   
5842   7.88 Lakh   4.75                 1                   0   

      Fuel_Type_Petrol  Transmission_Automatic  Transmission_Manual  
1                    1                       0                    1  
5                    0                       1                    0  
8                    1                       0                    1  
13                   0                       0                    1  
18                   0                       1                    0  
...                ...                     ...                  ...  
5827                 1                       0                    1  
5830                 1                       0                    1  
5833                 0                       0                    1  
5838                 1                       0                    1  
5842                 0                       0                    1  

[815 rows x 17 columns]

In [ ]:
df_cleaned['Car_Age'] = 2025 - df_cleaned['Year']
display(df_cleaned)

Unnamed: 0                               Name Location  Year  \
1              2                       Honda Jazz V  Chennai  2011   
5              7  Toyota Innova Crysta 2.8 GX AT 8S   Mumbai  2016   
8             10                   Maruti Ciaz Zeta    Kochi  2018   
13            15        Mitsubishi Pajero Sport 4X4    Delhi  2014   
18            20                  BMW 3 Series 320d    Kochi  2014   
...          ...                                ...      ...   ...   
5827        5999              Tata Bolt Revotron XT  Chennai  2016   
5830        6002      Volkswagen Vento 1.6 Highline   Mumbai  2011   
5833        6005           Maruti Vitara Brezza VDi     Pune  2016   
5838        6010               Honda Brio 1.2 VX MT    Delhi  2013   
5842        6014                   Maruti Swift VDI    Delhi  2014   

      Kilometers_Driven Owner_Type     Mileage   Engine       Power  Seats  \
1                 46000      First    13 km/kg  1199 CC    88.7 bhp    5.0   
5                 36000      First  11.36 kmpl  2755 CC   171.5 bhp    8.0   
8                 25692      First  21.56 kmpl  1462 CC  103.25 bhp    5.0   
13               110000      First   13.5 kmpl  2477 CC  175.56 bhp    7.0   
18                32982      First  22.69 kmpl  1995 CC     190 bhp    5.0   
...                 ...        ...         ...      ...         ...    ...   
5827              10000      First  17.57 kmpl  1193 CC    88.7 bhp    5.0   
5830              38000      First  16.09 kmpl  1598 CC   103.5 bhp    5.0   
5833              37208      First   24.3 kmpl  1248 CC    88.5 bhp    5.0   
5838              33746      First   18.5 kmpl  1198 CC    86.8 bhp    5.0   
5842              27365      First   28.4 kmpl  1248 CC      74 bhp    5.0   

       New_Price  Price  Fuel_Type_Diesel  Fuel_Type_Electric  \
1      8.61 Lakh   4.50                 0                   0   
5        21 Lakh  17.50                 1                   0   
8     10.65 Lakh   9.95                 0                   0   
13    32.01 Lakh  15.00                 1                   0   
18    47.87 Lakh  18.55                 1                   0   
...          ...    ...               ...                 ...   
5827   7.77 Lakh   4.00                 0                   0   
5830  11.91 Lakh   3.25                 0                   0   
5833   9.93 Lakh   7.43                 1                   0   
5838   6.63 Lakh   3.20                 0                   0   
5842   7.88 Lakh   4.75                 1                   0   

      Fuel_Type_Petrol  Transmission_Automatic  Transmission_Manual  Car_Age  
1                    1                       0                    1       14  
5                    0                       1                    0        9  
8                    1                       0                    1        7  
13                   0                       0                    1       11  
18                   0                       1                    0       11  
...                ...                     ...                  ...      ...  
5827                 1                       0                    1        9  
5830                 1                       0                    1       14  
5833                 0                       0                    1        9  
5838                 1                       0                    1       12  
5842                 0                       0                    1       11  

[815 rows x 18 columns]

dplyr in Python:

1) Select:

In [ ]:
selected_df = df_cleaned[['Name', 'Year']]
display(selected_df)

Name  Year
1                          Honda Jazz V  2011
5     Toyota Innova Crysta 2.8 GX AT 8S  2016
8                      Maruti Ciaz Zeta  2018
13          Mitsubishi Pajero Sport 4X4  2014
18                    BMW 3 Series 320d  2014
...                                 ...   ...
5827              Tata Bolt Revotron XT  2016
5830      Volkswagen Vento 1.6 Highline  2011
5833           Maruti Vitara Brezza VDi  2016
5838               Honda Brio 1.2 VX MT  2013
5842                   Maruti Swift VDI  2014

[815 rows x 2 columns]

2) Filter:

In [ ]:
selected_df = df_cleaned[(df_cleaned['Kilometers_Driven'] > 100000) & (df_cleaned['Price'] > 4) & (df_cleaned['Car_Age']  > 10)]
display(selected_df)

Unnamed: 0                         Name    Location  Year  \
13            15  Mitsubishi Pajero Sport 4X4       Delhi  2014   
535          555            Maruti Ertiga ZDI        Pune  2013   
553          573          Mahindra Bolero ZLX      Jaipur  2012   
913          946  Mitsubishi Pajero Sport 4X4   Bangalore  2013   
996         1030   Toyota Fortuner 2.8 4WD MT       Kochi  2013   
2178        2242   Toyota Fortuner 2.8 4WD MT   Hyderabad  2010   
2211        2277   Toyota Fortuner 2.8 4WD MT      Mumbai  2010   
2777        2866            BMW 3 Series 320d  Coimbatore  2009   
3525        3640   Toyota Fortuner 2.8 4WD MT       Kochi  2011   
3667        3786             Maruti Swift VDI        Pune  2014   
5105        5254       Hyundai Elantra 1.6 SX        Pune  2013   

      Kilometers_Driven Owner_Type     Mileage   Engine       Power  Seats  \
13               110000      First   13.5 kmpl  2477 CC  175.56 bhp    7.0   
535              105000     Second  25.47 kmpl  1248 CC   88.50 bhp    7.0   
553              105000     Second  15.96 kmpl  2523 CC    62.1 bhp    7.0   
913              132000      First   13.5 kmpl  2477 CC  175.56 bhp    7.0   
996              105072      First  14.24 kmpl  2755 CC   174.5 bhp    7.0   
2178             110000      First  14.24 kmpl  2755 CC   174.5 bhp    7.0   
2211             140000      First  14.24 kmpl  2755 CC   174.5 bhp    7.0   
2777             103905      First  22.69 kmpl  1995 CC     190 bhp    5.0   
3525             169006      First  14.24 kmpl  2755 CC   174.5 bhp    7.0   
3667             115000      First   28.4 kmpl  1248 CC      74 bhp    5.0   
5105             170000      First  22.54 kmpl  1582 CC   126.2 bhp    5.0   

       New_Price  Price  Fuel_Type_Diesel  Fuel_Type_Electric  \
13    32.01 Lakh  15.00                 1                   0   
535   12.34 Lakh   6.75                 1                   0   
553   11.15 Lakh   6.00                 1                   0   
913    34.6 Lakh  12.49                 1                   0   
996   36.69 Lakh  13.28                 1                   0   
2178  37.51 Lakh  13.70                 1                   0   
2211  37.46 Lakh  10.75                 1                   0   
2777  43.31 Lakh   9.92                 1                   0   
3525  36.69 Lakh  10.64                 1                   0   
3667   8.26 Lakh   4.50                 1                   0   
5105  20.87 Lakh   7.50                 1                   0   

      Fuel_Type_Petrol  Transmission_Automatic  Transmission_Manual  Car_Age  
13                   0                       0                    1       11  
535                  0                       0                    1       12  
553                  0                       0                    1       13  
913                  0                       0                    1       12  
996                  0                       0                    1       12  
2178                 0                       0                    1       15  
2211                 0                       0                    1       15  
2777                 0                       1                    0       16  
3525                 0                       0                    1       14  
3667                 0                       0                    1       11  
5105                 0                       0                    1       12

3) Rename:

In [ ]:
renamed_df = df_cleaned.rename(columns={'Name': 'Model_Name', 'Seats':'No. of Seats'})
display(renamed_df)

Unnamed: 0                         Model_Name Location  Year  \
1              2                       Honda Jazz V  Chennai  2011   
5              7  Toyota Innova Crysta 2.8 GX AT 8S   Mumbai  2016   
8             10                   Maruti Ciaz Zeta    Kochi  2018   
13            15        Mitsubishi Pajero Sport 4X4    Delhi  2014   
18            20                  BMW 3 Series 320d    Kochi  2014   
...          ...                                ...      ...   ...   
5827        5999              Tata Bolt Revotron XT  Chennai  2016   
5830        6002      Volkswagen Vento 1.6 Highline   Mumbai  2011   
5833        6005           Maruti Vitara Brezza VDi     Pune  2016   
5838        6010               Honda Brio 1.2 VX MT    Delhi  2013   
5842        6014                   Maruti Swift VDI    Delhi  2014   

      Kilometers_Driven Owner_Type     Mileage   Engine       Power  \
1                 46000      First    13 km/kg  1199 CC    88.7 bhp   
5                 36000      First  11.36 kmpl  2755 CC   171.5 bhp   
8                 25692      First  21.56 kmpl  1462 CC  103.25 bhp   
13               110000      First   13.5 kmpl  2477 CC  175.56 bhp   
18                32982      First  22.69 kmpl  1995 CC     190 bhp   
...                 ...        ...         ...      ...         ...   
5827              10000      First  17.57 kmpl  1193 CC    88.7 bhp   
5830              38000      First  16.09 kmpl  1598 CC   103.5 bhp   
5833              37208      First   24.3 kmpl  1248 CC    88.5 bhp   
5838              33746      First   18.5 kmpl  1198 CC    86.8 bhp   
5842              27365      First   28.4 kmpl  1248 CC      74 bhp   

      No. of Seats   New_Price  Price  Fuel_Type_Diesel  Fuel_Type_Electric  \
1              5.0   8.61 Lakh   4.50                 0                   0   
5              8.0     21 Lakh  17.50                 1                   0   
8              5.0  10.65 Lakh   9.95                 0                   0   
13             7.0  32.01 Lakh  15.00                 1                   0   
18             5.0  47.87 Lakh  18.55                 1                   0   
...            ...         ...    ...               ...                 ...   
5827           5.0   7.77 Lakh   4.00                 0                   0   
5830           5.0  11.91 Lakh   3.25                 0                   0   
5833           5.0   9.93 Lakh   7.43                 1                   0   
5838           5.0   6.63 Lakh   3.20                 0                   0   
5842           5.0   7.88 Lakh   4.75                 1                   0   

      Fuel_Type_Petrol  Transmission_Automatic  Transmission_Manual  Car_Age  
1                    1                       0                    1       14  
5                    0                       1                    0        9  
8                    1                       0                    1        7  
13                   0                       0                    1       11  
18                   0                       1                    0       11  
...                ...                     ...                  ...      ...  
5827                 1                       0                    1        9  
5830                 1                       0                    1       14  
5833                 0                       0                    1        9  
5838                 1                       0                    1       12  
5842                 0                       0                    1       11  

[815 rows x 18 columns]

4) Mutate

In [ ]:
df_2 = df_cleaned.copy()
df_2['Price_Per_Kilometer'] = df_2['Price']/df_2['Kilometers_Driven']
display(df_2)

Unnamed: 0                               Name Location  Year  \
1              2                       Honda Jazz V  Chennai  2011   
5              7  Toyota Innova Crysta 2.8 GX AT 8S   Mumbai  2016   
8             10                   Maruti Ciaz Zeta    Kochi  2018   
13            15        Mitsubishi Pajero Sport 4X4    Delhi  2014   
18            20                  BMW 3 Series 320d    Kochi  2014   
...          ...                                ...      ...   ...   
5827        5999              Tata Bolt Revotron XT  Chennai  2016   
5830        6002      Volkswagen Vento 1.6 Highline   Mumbai  2011   
5833        6005           Maruti Vitara Brezza VDi     Pune  2016   
5838        6010               Honda Brio 1.2 VX MT    Delhi  2013   
5842        6014                   Maruti Swift VDI    Delhi  2014   

      Kilometers_Driven Owner_Type     Mileage   Engine       Power  Seats  \
1                 46000      First    13 km/kg  1199 CC    88.7 bhp    5.0   
5                 36000      First  11.36 kmpl  2755 CC   171.5 bhp    8.0   
8                 25692      First  21.56 kmpl  1462 CC  103.25 bhp    5.0   
13               110000      First   13.5 kmpl  2477 CC  175.56 bhp    7.0   
18                32982      First  22.69 kmpl  1995 CC     190 bhp    5.0   
...                 ...        ...         ...      ...         ...    ...   
5827              10000      First  17.57 kmpl  1193 CC    88.7 bhp    5.0   
5830              38000      First  16.09 kmpl  1598 CC   103.5 bhp    5.0   
5833              37208      First   24.3 kmpl  1248 CC    88.5 bhp    5.0   
5838              33746      First   18.5 kmpl  1198 CC    86.8 bhp    5.0   
5842              27365      First   28.4 kmpl  1248 CC      74 bhp    5.0   

       New_Price  Price  Fuel_Type_Diesel  Fuel_Type_Electric  \
1      8.61 Lakh   4.50                 0                   0   
5        21 Lakh  17.50                 1                   0   
8     10.65 Lakh   9.95                 0                   0   
13    32.01 Lakh  15.00                 1                   0   
18    47.87 Lakh  18.55                 1                   0   
...          ...    ...               ...                 ...   
5827   7.77 Lakh   4.00                 0                   0   
5830  11.91 Lakh   3.25                 0                   0   
5833   9.93 Lakh   7.43                 1                   0   
5838   6.63 Lakh   3.20                 0                   0   
5842   7.88 Lakh   4.75                 1                   0   

      Fuel_Type_Petrol  Transmission_Automatic  Transmission_Manual  Car_Age  \
1                    1                       0                    1       14   
5                    0                       1                    0        9   
8                    1                       0                    1        7   
13                   0                       0                    1       11   
18                   0                       1                    0       11   
...                ...                     ...                  ...      ...   
5827                 1                       0                    1        9   
5830                 1                       0                    1       14   
5833                 0                       0                    1        9   
5838                 1                       0                    1       12   
5842                 0                       0                    1       11   

      Price_Per_Kilometer  
1                0.000098  
5                0.000486  
8                0.000387  
13               0.000136  
18               0.000562  
...                   ...  
5827             0.000400  
5830             0.000086  
5833             0.000200  
5838             0.000095  
5842             0.000174  

[815 rows x 19 columns]

5) Arrange


In [ ]:
arranged_df = df_cleaned.sort_values(by='Year', ascending=True)
display(arranged_df)

Unnamed: 0                                      Name    Location  Year  \
4723        4863                     Hyundai Santro D Lite     Chennai  2001   
5414        5569                     Honda CR-V Petrol 2WD     Chennai  2004   
337          351                          Skoda Octavia RS      Jaipur  2005   
2058        2119            Ford Freestyle Titanium Petrol      Mumbai  2007   
251          261                 Hyundai Verna CRDi 1.6 SX   Hyderabad  2008   
...          ...                                       ...         ...   ...   
226          235            Toyota Innova Crysta 2.8 GX AT       Kochi  2019   
1322        1368        Hyundai Grand i10 1.2 Kappa Sportz       Kochi  2019   
1281        1324                    Maruti Ignis 1.2 Delta  Coimbatore  2019   
65            67  Mercedes-Benz C-Class Progressive C 220d  Coimbatore  2019   
165          173          Hyundai Grand i10 1.2 Kappa Asta     Kolkata  2019   

      Kilometers_Driven Owner_Type     Mileage   Engine      Power  Seats  \
4723             140000      Third   20.3 kmpl  1086 CC     68 bhp    5.0   
5414             135000      Third   14.4 kmpl  1997 CC    152 bhp    5.0   
337              100000      First  14.45 kmpl  1984 CC    227 bhp    5.0   
2058               1261      First   19.0 kmpl  1194 CC  94.68 bhp    5.0   
251              143275      First   22.0 kmpl  1582 CC  126.2 bhp    5.0   
...                 ...        ...         ...      ...        ...    ...   
226               14165      First  11.36 kmpl  2755 CC  171.5 bhp    7.0   
1322              18453      First   18.9 kmpl  1197 CC  81.86 bhp    5.0   
1281              16862      First  20.89 kmpl  1197 CC  81.80 bhp    5.0   
65                15369      First    0.0 kmpl  1950 CC    194 bhp    5.0   
165                1000      First   18.9 kmpl  1197 CC  81.86 bhp    5.0   

       New_Price  Price  Fuel_Type_Diesel  Fuel_Type_Electric  \
4723   4.55 Lakh   0.70                 0                   0   
5414  33.51 Lakh   3.00                 0                   0   
337   28.37 Lakh   2.05                 0                   0   
2058   8.12 Lakh   6.00                 0                   0   
251   14.07 Lakh   2.60                 1                   0   
...          ...    ...               ...                 ...   
226   20.74 Lakh  19.64                 1                   0   
1322   6.68 Lakh   6.17                 0                   0   
1281   6.07 Lakh   6.06                 0                   0   
65    49.14 Lakh  35.67                 1                   0   
165    7.39 Lakh   6.50                 0                   0   

      Fuel_Type_Petrol  Transmission_Automatic  Transmission_Manual  Car_Age  
4723                 1                       0                    1       24  
5414                 1                       1                    0       21  
337                  1                       1                    0       20  
2058                 1                       0                    1       18  
251                  0                       0                    1       17  
...                ...                     ...                  ...      ...  
226                  0                       1                    0        6  
1322                 1                       0                    1        6  
1281                 1                       0                    1        6  
65                   0                       1                    0        6  
165                  1                       0                    1        6  

[815 rows x 18 columns]

6) Summarize with Group By:

In [ ]:
summarized_df = df_cleaned.groupby('Name')['Kilometers_Driven'].mean()
display(summarized_df)

Name
Audi A3 35 TDI Premium Plus           39000.000000
Audi A4 30 TFSI Premium Plus          61221.000000
Audi A4 35 TDI Premium Plus           33534.666667
Audi A4 35 TDI Technology             29500.000000
Audi A6 35 TDI Matrix                 43706.000000
                                          ...     
Volkswagen Vento 1.5 TDI Trendline    47000.000000
Volkswagen Vento 1.6 Comfortline      33576.500000
Volkswagen Vento 1.6 Highline         42143.666667
Volvo S60 D4 Momentum                 51000.000000
Volvo V40 D3 R Design                 38000.000000
Name: Kilometers_Driven, Length: 318, dtype: float64